pkg/storage/swift/swift.go

```diff
@@ -1,84 +1,294 @@
+	}
+
+	client, err := openstack.NewContainerV1(provider, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// NewDriver creates new Swift driver for the Image Registry
 func NewDriver(c *imageregistryv1.ImageRegistryConfigStorageSwift, listers *regopclient.Listers) *driver {
+	var secretLister kubelisters.SecretNamespaceLister
+	if c.CloudsSecret != nil {
+		kubeconfig, err := regopclient.GetConfig()
+		if err != nil {
+			// TODO(Fedosin): an error should also be returned in this case
+			return nil
+		}
+		kubeClient, err := kubeset.NewForConfig(kubeconfig)
+		if err != nil {
+			// TODO(Fedosin): ditto
+			return nil
+		}
+		kubeInformerFactory := kubeinformers.NewSharedInformerFactoryWithOptions(kubeClient, 10*time.Minute, kubeinformers.WithNamespace(c.CloudsSecret.Namespace))
+		secretLister = kubeInformerFactory.Core().V1().Secrets().Lister().Secrets(c.CloudsSecret.Namespace)
+	}
+
 	return &driver{
-		Config:  c,
-		Listers: listers,
+		Config:       c,
+		Listers:      listers,
+		SecretLister: secretLister,
 	}
 }
 
+// Secrets returns a map of the storage access secrets
 func (d *driver) Secrets() (map[string]string, error) {
+	if d.Config.CloudsSecret == nil {
+		cfg, err := clusterconfig.GetSwiftConfig(d.Listers)
+		if err != nil {
+			return nil, err
+		}
+
+		return map[string]string{
+			"REGISTRY_STORAGE_SWIFT_USERNAME": cfg.Storage.Swift.Username,
+			"REGISTRY_STORAGE_SWIFT_PASSWORD": cfg.Storage.Swift.Password,
+		}, nil
+	}
+
 	return nil, nil
 }
 
 func (d *driver) ConfigEnv() (envs []corev1.EnvVar, err error) {
+	// Add common values that do not depend on the config source
 	envs = append(envs,
 		corev1.EnvVar{Name: "REGISTRY_STORAGE", Value: "swift"},
-		corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_AUTHURL", Value: d.Config.AuthURL},
 		corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_CONTAINER", Value: d.Config.Container},
-		corev1.EnvVar{
-			Name: "REGISTRY_STORAGE_SWIFT_USERNAME",
-			ValueFrom: &corev1.EnvVarSource{
-				SecretKeyRef: &corev1.SecretKeySelector{
-					LocalObjectReference: corev1.LocalObjectReference{
-						Name: imageregistryv1.ImageRegistryPrivateConfiguration,
+	)
+	if d.Config.CloudsSecret != nil {
+		// If these values are provided read remaining values from the cloud config
+		cloud, err := d.getCloudFromSecret(d.Config.CloudsSecret.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		envs = append(envs,
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_AUTHURL", Value: cloud.AuthInfo.AuthURL},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_DOMAIN", Value: cloud.AuthInfo.DomainName},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_DOMAINID", Value: cloud.AuthInfo.DomainID},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_TENANT", Value: cloud.AuthInfo.ProjectName},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_TENANTID", Value: cloud.AuthInfo.ProjectID},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_USERNAME", Value: cloud.AuthInfo.Username},
```

Comment:
You shouldn't expose secrets there. That's the reason why we use EnvVarSource from ImageRegistryPrivateConfiguration.